### PR TITLE
Clean up the Lezer grammar output tree

### PR DESCRIPTION
**Pull request by marijnh**
I have been asked to review the `lezer-promql` grammar, and noticed that it was creating more nodes and a more complex tree structure than it needed to. The first patch in this PR makes the following changes:

 - Removes the `Expr` wrapping node around expressions, tagging the (lowercase) `expr` rule with `@isGroup=Expr` so that you can still query whether something is an expression.
 - Replaces the `MetricIdentifier(Identifier)` structure with just `Identifier`.
 - Makes the binary op modifiers optional, and splits `BoolModifier` from ignoring/on modifiers (`ModifierClause`) for clarity.
 - Collapses recursive rules (`GroupingLabelList`, `FunctionCallArgs`, `LabelMatchList`) that just existed to implement a repetition into a simpler rule that uses the `*` operator and creates a flat tree.

I've adjusted the `lezer-promql` tests to test for the updated tree shape.

However, after doing this I found out the `codemirror-promql` package had some code that is very tightly coupled to the way the syntax tree looks. I've tried to adjust this in the second patch, **but** I was unable to run the test suite, so this is most likely not done. Firstly, the package files for the packages under module/ seem to miss some dev dependencies (jest, rollup) and rely on higher-up packages to also be installed. After doing that, I could start jest, but it still failed to run with messages like...

> ReferenceError: exports is not defined

> The 'import.meta' meta-property is only allowed when the '--module' option is 'es2020', 'es2022', 'esnext', 'system', 'node16', or 'nodenext'.
https://github.com/prometheus/prometheus/pull/11333
... suggesting some kind of module-system confusion. (I'm using Node 16.13.) Would anybody be able to point me in the right direction regarding how to run these tests?

### CODE REVIEW

**Comment by juliusv** [comment]
Awesome, thanks! I haven't studied the code changes yet (aim to do that soon), but for the tests, this is what works for me:

```bash
$ node --version
v16.16.0
$ npm --version
8.11.0

$ cd web/ui
$ npm i
$ npm test # Run tests for all the web stuff
$ cd module/codemirror-promql
$ npm test # Run tests just for codemirror-promql
```

File: web/ui/module/lezer-promql/src/promql.grammar [hunk 1]
```diff
@@ -65,53 +66,38 @@ AggregateModifier {
...
+  expr !eql Neq    binModifiers expr |
+  expr !and And    binModifiers expr |
+  expr !and Unless binModifiers expr |
+  expr !or  Or     binModifiers expr
 }
 
-OnOrIgnoring {
-  Ignoring GroupingLabels |
-  On GroupingLabels
+ModifierClause {
```
[the remark this answers is not in the document]
Comment (marijnh) [reply web/ui/module/lezer-promql/src/promql.grammar]: Sure! I wasn't really certain what this did, which somewhat impaired my ability to name it. I've updated the branch to rename this to `MatchingModifierClause`.

**Comment by juliusv** [comment]
Perfect, thanks! I'll merge this in the next few days then, unless @Nexucis has any complaints during that time.

**Comment by juliusv** [comment]
Oops, we're actually getting one linter error now (just some code formatting):

```
/__w/prometheus/prometheus/web/ui/module/codemirror-promql/src/parser/vector.ts
  16:9  error  Replace `·And,·BinaryExpr,·MatchingModifierClause,·LabelName,·GroupingLabels,·GroupLeft,·GroupRight,·On,·Or,·Unless·` with `⏎··And,⏎··BinaryExpr,⏎··MatchingModifierClause,⏎··LabelName,⏎··GroupingLabels,⏎··GroupLeft,⏎··GroupRight,⏎··On,⏎··Or,⏎··Unless,⏎`  prettier/prettier

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
```

File: web/ui/module/codemirror-promql/src/parser/parser.ts [hunk 2]
```diff
@@ -117,28 +110,28 @@ export class Parser {
...
       case MatrixSelector:
-        this.checkAST(walkThrough(node, Expr));
+        this.checkAST(node.getChild('Expr'));
         break;
       case VectorSelector:
         this.checkVectorSelector(node);
         break;
       case StepInvariantExpr:
-        const exprValue = this.checkAST(walkThrough(node, Expr));
+        const exprValue = this.checkAST(node.getChild('Expr'));
```
Comment (Nexucis) [inline web/ui/module/codemirror-promql/src/parser/parser.ts:134]: isn't possible to pass the node ID instead ? 

I would prefer to keep the usage of the constant `Expr`

**Review by Nexucis — COMMENTED** [review]
impressive changes there. Thank you for doing that @marijnh ! I did not think `node.getChild` was able to replace the util method `walkBackward`.

I just put a comment as a question, but otherall, yup looks much cleaner than before and the grammar is simpler. Amazing !

File: web/ui/module/codemirror-promql/src/parser/parser.ts (hunk 2 above)
Comment (marijnh) [reply web/ui/module/codemirror-promql/src/parser/parser.ts:134]: Node groups (see [here](https://lezer.codemirror.net/docs/ref/#common.NodeProp^group)) are identified only by strings.

**Comment by marijnh** [comment]
Updated the PR with the whitespace change.

File: web/ui/module/codemirror-promql/src/parser/parser.ts (hunk 2 above)
Comment (Nexucis) [reply web/ui/module/codemirror-promql/src/parser/parser.ts:134]: ah yes I didn't pay enough attention to the changes in the grammar.
Thank you for the explanation :)

**Review by Nexucis — APPROVED** [review]
LGTM

**Comment by gtk-grafana** [comment]
@marijnh This change breaks Grafana's current usage of this module, we are currently using `BinModifiers`, `OnOrIgnoring`, `LabelMatchList`, `MetricIdentifier`, `Expr`, `FunctionCallArgs`, `GroupingLabel`, and `GroupingLabelList` which no longer exist as of this merge. Some usages appear to have trivial replacements, but for others it is not clear to me how we should be migrating to support this change.

The unit tests are helpful for seeing how we should update, but most of the previously exported variables above are not covered.

Are there any tips or documentation you can provide for replacing the usages of these exports that are no longer available?

**Comment by juliusv** [comment]
@gtk-grafana I'm sorry to hear! FWIW, Grafana has started using this module for purposes that it wasn't intended or suited for, and to my knowledge, without consulting with the original developers (myself and @Nexucis) first. Generally I'm not sure if that's a good idea, as lezer trees are optimized towards CodeMirror-like text-editing use cases (optimizing for things like parse error recovery, incremental parsing, etc.), but not as great for doing other high-level interpretation and manipulation of symbolic ASTs, like I understand you're currently using it for in the graphical query builder in Grafana.

See also this excerpt from https://lezer.codemirror.net/docs/guide/:

> These trees, represented by data structures from the [@lezer/common](https://lezer.codemirror.net/docs/ref/#common) package, are more limited than the abstract syntax trees you might have seen in other contexts. They are not very abstract. Each node only stores a type, a start and end position, and a flat sequence of children. When writing a grammar, you choose which productions are stored as nodes—the others don't appear in the tree at all.

Now given the current reality of Grafana using this (so far, successfully) for the query editor, I'm not sure exactly how to migrate things there, but I hope it's possible of course. But I wanted to clarify that this is not generally a use case we were looking to explicitly support with this package.

**Pull request merged**